### PR TITLE
Flow changes which allow set of instances to bootstrap and other changes

### PR DESCRIPTION
**Pull request by anirbanr-fb**
Summary:

1. distributed config. Once you provide tserver addresses, the peers are
created and the config with bootstrap tserver is put onto disk for next time

2. The above allows re-entrancy. The next time the bootstrap config is already
in place and will be picked up by raft.

3. Continuing the set of changes that ritwik started by making the metrics as
server instead of tablet.Maybe we should roll this back in the future to stay
as close to upstream as possible.

4. Bring back Kerberos and messenger intializations so that a set of nodes can
talk to each other.

5. Fixing the header file collision issue with sparsehash and kudu/gutil

Test Plan: Ran unittests and all key unittests succeeded.

Brough up in fbcode a set of heartbeating servers who successfully elected
leader and did leader elections on failure. No kerberos and TLS issues were
encountered as kudu worked fine with self-signing certificates

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: src/kudu/gutil/template_util.h [hunk 1]
```diff
@@ -46,8 +46,10 @@
 // Both of these outcomes means that we may be able to directly replace
 // some of these with boost equivalents.
 //
-#ifndef BASE_TEMPLATE_UTIL_H_
-#define BASE_TEMPLATE_UTIL_H_
+
+// These #defines collide with sparsehash
+#ifndef KUDU_BASE_TEMPLATE_UTIL_H_
```
Comment (iRitwik) [inline src/kudu/gutil/template_util.h:51]: This is a clever mitigation :) Will this stay for good or is it temporary setup for milestone 1?

File: src/kudu/server/server_base.cc [hunk 2]
```diff
@@ -128,7 +128,6 @@ DEFINE_string(user_acl, "*",
 TAG_FLAG(user_acl, stable);
 TAG_FLAG(user_acl, sensitive);
 
-#ifdef FB_DO_NOT_REMOVE
 DEFINE_string(principal, "kudu/_HOST",
```
Comment (iRitwik) [inline src/kudu/server/server_base.cc:131]: This is necessary for x-DC traffic. I am not sure how we handle this with replication right now. Is similar logic present in MySQL replication plugin to encrypt and send traffic? In the long term, should we aim to consolidate the logic in one place?

We also need to hardcode the minimum TLS version to 1.2 in order to maintain compliance but that would be trivial.

File: src/kudu/tserver/tablet_server_options.cc [hunk 3]
```diff
@@ -29,32 +29,45 @@
...
+    if (tserver_addresses.size() == 2) {
+      LOG(WARNING) << "Only 2 tservers are specified by tserver_addresses_flag ('" <<
+          FLAGS_tserver_addresses << "'), but minimum of 3 are required to tolerate failures"
+          " of any one tserver. It is recommended to use at least 3 tservers.";
+    }
   }
-#endif
+}
+
+bool TabletServerOptions::IsDistributed() const {
```
Comment (iRitwik) [inline src/kudu/tserver/tablet_server_options.cc:69]: Would it matter if more servers join the raft ring a little later via a config change and this function would give us stale information?

nit: `isStandaloneNode` might be a more indicative name for the function.

File: src/kudu/tserver/tablet_server.cc [hunk 4]
```diff
@@ -129,12 +130,35 @@ Status TabletServer::Init() {
...
   return Status::OK();
-  //return tablet_manager_->WaitForAllBootstrapsToFinish();
+}
+
+Status TabletServer::WaitForTabletManagerInit() const {
+  return init_status_.Get();
 }
 
 Status TabletServer::Start() {
+  RETURN_NOT_OK(StartAsync());
```
Comment (iRitwik) [inline src/kudu/tserver/tablet_server.cc:152]: Why did we asyncify the initialization?

File: src/kudu/tserver/simple_tablet_manager.cc [hunk 5]
```diff
@@ -160,6 +337,16 @@ Status TSTabletManager::Init() {
   VLOG(2) << "RaftConfig before starting: " << SecureDebugString(consensus_->CommittedConfig());
 
   scoped_refptr<Log> log;
+
+  OpId init;
+  init.set_term(0);
+  init.set_index(0);
+  scoped_refptr<MetricEntity> metric_entity;
+  metric_entity = METRIC_ENTITY_server.Instantiate(metric_registry_, kSysCatalogTabletId);
```
Comment (iRitwik) [inline src/kudu/tserver/simple_tablet_manager.cc]: Why can't we re-use the metric entity of the server?

File: src/kudu/tserver/simple_tablet_manager.cc [hunk 6]
```diff
@@ -117,36 +122,208 @@ TSTabletManager::TSTabletManager(TabletServer* server)
...
+  }
+
+  // Now resolve UUIDs.
+  // By the time a SysCatalogTable is created and initted, the masters should be
+  // starting up, so this should be fine to do.
+  DCHECK(server_->messenger());
+  RaftConfigPB resolved_config = new_config;
+  resolved_config.clear_peers();
+  for (const RaftPeerPB& peer : new_config.peers()) {
+    if (peer.has_permanent_uuid()) {
```
Comment (iRitwik) [inline src/kudu/tserver/simple_tablet_manager.cc:217]: I don't think the if block would ever execute because we haven't set the uuid in the protobufs in 201-208.

File: src/kudu/tserver/simple_tablet_manager.cc [hunk 7]
```diff
@@ -117,36 +122,208 @@ TSTabletManager::TSTabletManager(TabletServer* server)
...
+  }
+
+  RETURN_NOT_OK(consensus::VerifyRaftConfig(resolved_config));
+  VLOG(1) << "Distributed Raft configuration: " << SecureShortDebugString(resolved_config);
+
+  *committed_config = resolved_config;
+  return Status::OK();
+}
+
+Status TSTabletManager::WaitUntilConsensusRunning(const MonoDelta& timeout) {
```
Comment (iRitwik) [inline src/kudu/tserver/simple_tablet_manager.cc:238]: timeout should prolly be a flag

File: src/kudu/tserver/simple_tablet_manager.cc [hunk 8]
```diff
@@ -117,36 +122,208 @@ TSTabletManager::TSTabletManager(TabletServer* server)
...
+Status TSTabletManager::Init(bool is_first_run) {
   CHECK_EQ(state(), MANAGER_INITIALIZING);
 
-  InitLocalRaftPeerPB();
+  RETURN_NOT_OK_PREPEND(InitRaftAsync(is_first_run),
+                        "Failed to initialize raft async");
 
-  LOG(INFO) << LogPrefix(kSysCatalogTabletId) << "Bootstrapping tablet";
-  TRACE("Bootstrapping tablet");
+  // WaitUntilRunning() must run outside of the lock as to prevent
```
Comment (iRitwik) [inline src/kudu/tserver/simple_tablet_manager.cc]: Which lock are you referring to here? I got a bit confused.

File: src/kudu/tserver/simple_tablet_manager.h [hunk 9]
```diff
@@ -122,13 +130,27 @@ class TSTabletManager : public consensus::ConsensusRoundHandler {
...
 
+  Status WaitUntilRunning();
+
+  Status WaitUntilConsensusRunning(const MonoDelta& timeout);
+
+  Status CreateNew(FsManager *fs_manager);
+
+  Status SetupRaft();
+
+  Status InitRaftAsync(bool is_first_run);
```
Comment (iRitwik) [inline src/kudu/tserver/simple_tablet_manager.h:145]: nit: Most of them seem pretty straightforward after reading the cc file. Maybe leave one-liners in the headers to briefly explain what the member functions do. I have this habit of reading only headers. Upto you though.

**Review by iRitwik — COMMENTED** [review]
Overall, the change looks great. I have left a few comments to verify a few things post which I can accept. 

3. I understand the intent. However, I think at this point merging from upstream would be painful regardless which way we go. We might as well start perfecting our RPC server implementation to be exposed to the plugin. What do you think?

4. I am reading through the messenger implementation. Haven't finished yet. Is there a hard dependency on catalog manager?

File: src/kudu/gutil/template_util.h (hunk 1 above)
Comment (toddlipcon) [reply src/kudu/gutil/template_util.h:51]: Happy to fix this upstream to use pragma once

File: src/kudu/tserver/simple_tablet_manager.h [hunk 10]
```diff
@@ -122,13 +130,27 @@ class TSTabletManager : public consensus::ConsensusRoundHandler {
...
 
   // Initializes the RaftPeerPB for the local peer.
   // Guaranteed to include both uuid and last_seen_addr fields.
   // Crashes with an invariant check if the RPC server is not currently in a
   // running state.
   void InitLocalRaftPeerPB();
 
+  // Use the master options to generate a new consensus configuration.
+  // In addition, resolve all UUIDs of this consensus configuration.
+  Status CreateDistributedConfig(const TabletServerOptions& options,
```
Comment (bhatvinay) [inline src/kudu/tserver/simple_tablet_manager.h:155]: What is a distributed config? Also, does resolving here mean trying to communicate explicitly with the peer having this uuid?

File: src/kudu/tserver/simple_tablet_manager.cc [hunk 11]
```diff
@@ -160,6 +337,16 @@ Status TSTabletManager::Init() {
   VLOG(2) << "RaftConfig before starting: " << SecureDebugString(consensus_->CommittedConfig());
 
   scoped_refptr<Log> log;
+
+  OpId init;
+  init.set_term(0);
+  init.set_index(0);
```
Comment (bhatvinay) [inline src/kudu/tserver/simple_tablet_manager.cc]: Why is this needed?

File: src/kudu/gutil/template_util.h (hunk 1 above)
Comment (iRitwik) [reply src/kudu/gutil/template_util.h:51]: That would be great, @toddlipcon

File: src/kudu/tserver/simple_tablet_manager.cc (hunk 6 above)
Comment (anirbanr-fb) [reply src/kudu/tserver/simple_tablet_manager.cc:217]: yes, you are right. I think this was kept just for safety in the original code.

File: src/kudu/tserver/simple_tablet_manager.h (hunk 10 above)
Comment (anirbanr-fb) [reply src/kudu/tserver/simple_tablet_manager.h:155]: distributed config is a multi instance config. 
Resolve means contacting the host-ports to get their UUID.

File: src/kudu/tserver/simple_tablet_manager.h (hunk 9 above)
Comment (anirbanr-fb) [reply src/kudu/tserver/simple_tablet_manager.h:145]: Yes, we should document these functions.

File: src/kudu/tserver/simple_tablet_manager.cc (hunk 11 above)
Comment (anirbanr-fb) [reply src/kudu/tserver/simple_tablet_manager.cc]: Seems like this was not required, but even kudu had this dead code.

File: src/kudu/tserver/simple_tablet_manager.cc (hunk 5 above)
Comment (anirbanr-fb) [reply src/kudu/tserver/simple_tablet_manager.cc]: I think this was confusion between  METRIC_ENTITY_tablet and METRIC_ENTITY_server.

File: src/kudu/tserver/tablet_server_options.cc (hunk 3 above)
Comment (anirbanr-fb) [reply src/kudu/tserver/tablet_server_options.cc:69]: Distributed is the opposite of standalone. This is what kudu also uses.
There is an assertion which checks that min # of servers is 2

Comment (anirbanr-fb) [reply src/kudu/tserver/tablet_server_options.cc:69]: This option is only used in the very beginning of raft initialization to create the initial  fs and config, or validate the last config. So it is expected that this will go stale with later state of the ring.

File: src/kudu/tserver/simple_tablet_manager.cc (hunk 7 above)
Comment (anirbanr-fb) [reply src/kudu/tserver/simple_tablet_manager.cc:238]: this timeout is the amount of timeout a single call to WaitUntilConsensusRunning is supposed to wait.
At this moment, the higher level WaitUntilRunning waits for ever or error. We will need to revisit this soon.
We should make this a flag in WaitUntilRunning

File: src/kudu/tserver/tablet_server.cc (hunk 4 above)
Comment (anirbanr-fb) [reply src/kudu/tserver/tablet_server.cc:152]: Just followed the naming convention inside Kudu. The start function was always submitting the closure to init_pool.

File: src/kudu/tserver/simple_tablet_manager.cc (hunk 8 above)
Comment (anirbanr-fb) [reply src/kudu/tserver/simple_tablet_manager.cc]: This comment is from kudu. I had left it as a reminder of concurrency issues which might be
TBD, but I will remove it to avoid confusion.

File: src/kudu/server/server_base.cc (hunk 2 above)
Comment (anirbanr-fb) [reply src/kudu/server/server_base.cc:131]: brought it back.

File: src/kudu/tserver/simple_tablet_manager.cc (hunk 7 above)
Comment (anirbanr-fb) [reply src/kudu/tserver/simple_tablet_manager.cc:238]: left flag addition for future if necessary.

**Review by iRitwik — APPROVED** [review]
